core/src/main/scala/org/apache/spark/scheduler/ExecutorLossReason.scala

```diff
@@ -23,13 +23,29 @@ import org.apache.spark.executor.ExecutorExitCode
  * Represents an explanation for a executor or whole slave failing or exiting.
  */
 private[spark]
-class ExecutorLossReason(val message: String) {
+class ExecutorLossReason(val message: String) extends Serializable {
   override def toString: String = message
 }
 
+private[spark] case class ExecutorExitedAbnormally(val exitCode: Int, reason: String)
+  extends ExecutorLossReason(reason) {
+}
+
+private[spark] object ExecutorExitedAbnormally {
+  def apply(exitCode: Int): ExecutorExitedAbnormally = {
+    ExecutorExitedAbnormally(exitCode, ExecutorExitCode.explainExitCode(exitCode))
+  }
+}
+
 private[spark]
-case class ExecutorExited(val exitCode: Int)
-  extends ExecutorLossReason(ExecutorExitCode.explainExitCode(exitCode)) {
+case class ExecutorExitedNormally(val exitCode: Int, reason: String)
+  extends ExecutorLossReason(reason) {
+}
+
+private[spark] object ExecutorExitedNormally {
```

Comment:
I don't know, I find `ExecutorExitedAbnormally` and `ExecutorExitedNormally` a little confusing, since internally they hold exactly the same data (even the same reason message). What if there was only `ExecutorExited` with a parameter saying whether it should be treated as an error or not?